pkg/dhclient/dhclient.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2013-2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package dhclient
+
+import (
+	"context"
+	"log"
+	"regexp"
+	"time"
+
+	"github.com/u-root/u-root/pkg/dhclient"
+	"github.com/vishvananda/netlink"
+)
+
+// Request sets up the dhcp configurations for all of the ifNames.
+func Request(ifName string, timeout int, retry int, verbose bool, ipv4 bool, ipv6 bool) {
+	ifRE := regexp.MustCompilePOSIX(ifName)
+
+	ifnames, err := netlink.LinkList()
+	if err != nil {
+		log.Fatalf("Can't get list of link names: %v", err)
+	}
+
+	var filteredIfs []netlink.Link
+	for _, iface := range ifnames {
+		if ifRE.MatchString(iface.Attrs().Name) {
+			filteredIfs = append(filteredIfs, iface)
+		}
+	}
+
+	if len(filteredIfs) == 0 {
+		log.Fatalf("No interfaces match %s", ifName)
+	}
+
+	channel := make(chan error)
+	go configureAll(filteredIfs, channel, timeout, retry, verbose, ipv4, ipv6)
+	x := <-channel
+	log.Printf("The channel returns: %v", x)
+}
+
+func configureAll(ifs []netlink.Link, channel chan error, timeout int, retry int, verbose bool, ipv4 bool, ipv6 bool) {
+	packetTimeout := time.Duration(timeout) * time.Second
+
+	ctx, cancel := context.WithTimeout(context.Background(), packetTimeout*time.Duration(1<<uint(retry)))
+	defer cancel()
+
+	c := dhclient.Config{
+		Timeout: packetTimeout,
+		Retries: retry,
+	}
+	if verbose {
+		c.LogLevel = dhclient.LogSummary
+	}
+	r := dhclient.SendRequests(ctx, ifs, ipv4, ipv6, c)
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Printf("Done with dhclient: %v", ctx.Err())
+			return
+
+		case result, ok := <-r:
+			if !ok {
+				log.Printf("Configured all interfaces.")
+				return
```

Comment:
Close the channel before returning. You can do a "defer close(channel)" near the start of this function to close it on all return paths.

Reply:
Oh, yes this is a good idea. As of right now, the channel is still open after having returned out of the function. I will make sure to close it. Thanks for the catch :)!